polars/polars-io/src/csv/write_impl.rs

```diff
@@ -165,13 +165,35 @@ pub(crate) fn write<W: Write>(
     writer: &mut W,
     df: &DataFrame,
     chunk_size: usize,
-    options: &SerializeOptions,
+    options: &mut SerializeOptions,
 ) -> Result<()> {
     // check that the double quote is valid utf8
     std::str::from_utf8(&[options.quote, options.quote])
         .map_err(|_| PolarsError::ComputeError("quote char leads invalid utf8".into()))?;
     let delimiter = char::from(options.delimiter);
 
+    // if datetime format not specified, infer the maximum required precision
+    if options.datetime_format.is_none() {
+        for col in df.get_columns() {
+            match col.dtype() {
+                DataType::Datetime(TimeUnit::Microseconds, _)
+                    if options.datetime_format.is_none() =>
+                {
+                    options.datetime_format = Some("%FT%H:%M:%S.%6f".to_string());
+                }
+                DataType::Datetime(TimeUnit::Nanoseconds, _) => {
+                    options.datetime_format = Some("%FT%H:%M:%S.%9f".to_string());
+                    break; // highest precision; no need to check further
+                }
+                _ => {}
+            }
+        }
+        // if still not set, no cols require higher precision than "ms" (or no datetime cols)
+        if options.datetime_format.is_none() {
+            options.datetime_format = Some("%FT%H:%M:%S.%3f".to_string());
+        }
+    }
```

Comment:
@alexander-beedie you mention that your approach is a compromise since per-column formatting was too heavyweight. Accommodating per-column formatting would be pretty surgical, though -- you could simply change [these](https://github.com/pola-rs/polars/blob/8cd4b0b76d30d6496691755c097ac6d750f3520f/polars/polars-io/src/csv/write_impl.rs#L86-L96) lines to

```rust
            None => {
                let dt = match tu {
                    TimeUnit::Nanoseconds => temporal_conversions::timestamp_ns_to_datetime(v),
                    TimeUnit::Microseconds => temporal_conversions::timestamp_us_to_datetime(v),
                    TimeUnit::Milliseconds => temporal_conversions::timestamp_ms_to_datetime(v),
                };
                let fmt = match &options.datetime_format {
                    None => match tu {
                        TimeUnit::Nanoseconds => "%FT%H:%M:%S.%9f".to_string(),
                        TimeUnit::Microseconds => "%FT%H:%M:%S.%6f".to_string(),
                        TimeUnit::Milliseconds => "%FT%H:%M:%S.%3f".to_string(),
                    },
                    Some(fmt) => fmt,
                }
                write!(f, "{}", dt.format(fmt))
            }
```

Are you afraid of a performance penalty this way?

Reply:
Actually, I think this would be faster than your way -- primarily because it requires no additional looping

Reply:
This would be a heap alloc to string every value written. That would be slow. 

I don't think we need the string alloc. But I don't think we can make this performance claims, as this hits every written value.

Reply:
Ah right, this is on a per-value basis, not per-column. Fair enough

Reply:
You are right -- the `.to_string()` bit is unnecessary. Looking into the performance a bit deeper, there doesn't seem to be much of a penalty with my implementation (unless 0.7ms is considered significant).

```python
df = pl.Series(pd.date_range("2000-01-01", "2100-01-01")).to_frame()

%timeit df.write_csv()

# my suggestion
# 45.2 ms ± 579 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

# master 
# 44.5 ms ± 356 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Reply:
You are right -- the `.to_string()` bit is unnecessary.

There isn't much of a performance penalty to my approach

```python
df = pl.Series(pd.date_range("2000-01-01", "2100-01-01")).to_frame()

%timeit df.write_csv()

# master
46.5 ms ± 1.38 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

# my approach
46.5 ms ± 777 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Reply:
This is not really a good benchmark. On writing such a small DataFrame, we are only benchmarking virtual dispatch, and likely python interop. Try measuring different row sizes up to a millions rows or so, such that we benchmark the hot loop, not the setup code.

This different row sizes tells us more about how it deals with caches.

Reply:
@matteosantama
> you mention that your approach is a compromise since per-column formatting was too heavyweight

Mostly because on my first approach I found myself making it unnecessarily complicated (I was allowing a different user-specified format per time-unit and it became... unruly ;). So, I pulled back to a cleaner implementation that would slide into the existing code with minimal fuss.

@ritchie46 / @matteosantama 
>  Try measuring different row sizes up to a millions rows or so, such that we benchmark the hot loop, not the setup

Done; the per-time unit approach above becomes ~5-7% slower on larger data (I was trying on ~63,000,000 rows, as created by: `df = pl.Series(pl.date_range(date(2000,1,1), date(2001,12,31), "1s")).to_frame()`). So not _much_ slower, but not faster - at least, not on my home setup (vanilla M1).

Reply:
Yep, I'd expect something like that.

I think we should go for your approach. It is simple.

We can explore something more fine grained later, but I think that should require some effort to not loose performance.

Reply:
@alexander-beedie thanks for checking!